website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md

```diff
@@ -0,0 +1,412 @@
+
+## Cost of different syscall paths
+
+[Defense-in-depth](https://gvisor.dev/blog/2019/11/18/gvisor-security-basics-part-1/#defense-in-depth)
+is the key design principle of gVisor. In gVisor, different syscalls have
+different paths, further leading to different cost (orders of magnitude) on
+latency and CPU consumption. Here are the syscall paths in gVisor.
+
+![Figure 1](/assets/images/2021-11-11-syscall-figure1.png "Sentry syscall paths.")
+
+### Path 1: User-space vDSO
+
+Sentry provides [vDSO library](https://github.com/google/gvisor/tree/master/vdso)
+for its sandboxed processes. Several syscalls are short circuited and
+implemented in user space. These syscalls cost almost as much as native Linux.
+But note that the vDSO library is partially implemented. We once noticed some
+[syscalls](https://github.com/google/gvisor/issues/3101) in our environment are
+not properly terminated in user space. We create some additional implementations
+to the vDSO, and aim to push these improvements upstream when possible.
+
+### Path 2: Sentry contained
+
+Most syscalls, e.g., <code>clone(2)</code>, are implemented in Sentry. They are
+some basic abstractions of a operating system, such as process/thread lifecycle,
+scheduling, IPC, memory management, etc. These syscalls and all below suffer
+from a structural cost of syscall interception. The overhead is about 800ns
+while that of the native syscalls is about 70ns. We'll dig it further below.
+Syscalls of this kind spend takes about several microseconds, which is
+competitive to the corresponding native Linux syscalls.
+
+### Path 3: Host-kernel involved
+
+Some syscalls, resource related, e.g., read/write, are redirected into the host
+kernel. Note that gVisor never passes through application syscalls directly into
+host kernel for functional and security reasons. So comparing to native Linux,
+the time spent in Sentry seems an extra overhead. Another overhead is the way
+to call host kernel syscall. Let's use kvm platform of x86_64 as an example.
+After Sentry issues the syscall instruction, if it is in GR0, it firstly goes
+to the syscall entrypoint defined in LSTAR, and then halts to HR3 (a vmexit
+happens here), and exits from a signal handler, and executes syscall instruction
+again. We can save the "Halt to HR3" by introducing vmcall here, but there's
+still a syscall trampoline there and the vmexit/vmentry overhead is not trivial.
+Nevertheless, these overhead is not that significant.
+
+For some sentry-contained syscalls in Path 2, although the syscall semantic is
+terminated in Sentry, it may further introduces one or many unexpected exits to
+host kernel. It could be a page fault when Sentry runs, and more likely, a
+schedule event in Go runtime, e.g., M idle/wakeup. An example in hand is that
+<code>futex(FUETX_WAIT)</code> and <code>epoll_wait(2)</code> could lead to M
+idle and a further futex call into host kernel if it does not find any runnable
+Gs.
```

Comment:
Add a link for an explanation.

"(See the comments in https://go.dev/src/runtime/proc.go for further explanation about the GMP scheduler)."